File: backend/src/main/java/com/petconnect/backend/user/application/service/UserService.java
package com.petconnect.backend.user.application.service;

import com.petconnect.backend.user.application.dto.*;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.lang.Nullable;
import java.io.IOException;

import java.util.Optional;

/**
 * Service interface for managing general UserEntity operations, including retrieving profiles,
 * finding users, and potentially updating common profile information.
 * Specific role operations (Owner registration, Clinic Staff creation/management)
 * might be handled in more specialized services (AuthService, ClinicStaffService).
 *
 * @author ibosquet
 */
public interface UserService {

    /**
     * Loads user-specific data. This method is required by the UserDetailsService interface.
     * It uses the username (which could be email in our case) to find the user.
     *
     * @param username The username (or email) identifying the user whose data is required.
     * @return a fully populated {@link UserDetails} object (never {@code null})
     * @throws UsernameNotFoundException if the user could not be found or the user has no
     *                                   GrantedAuthority
     */
    UserDetails loadUserByUsername(String username) throws UsernameNotFoundException ;

    /**
     * Retrieves the profile information of the currently authenticated user.
     * Determines the user type and returns the appropriate specific profile DTO.
     *
     * @return A profile DTO (OwnerProfileDto or ClinicStaffProfileDto) cast to Object,
     *         or consider a common Profile interface if preferred. Returns null if no user is found.
     * @throws IllegalStateException if there is no authenticated user in the security context.
     */
    Object getCurrentUserProfile();

    /**
     * Finds a user by their unique identifier.
     *
     * @param id The ID of the user.
     * @return An Optional containing the generic UserProfileDto if found, otherwise empty.
     */
    Optional<UserProfileDto> findUserById(Long id);

    /**
     * Finds a user by their unique email address.
     *
     * @param email The email address.
     * @return An Optional containing the generic UserProfileDto if found, otherwise empty.
     */
    Optional<UserProfileDto> findUserByEmail(String email);

    /**
     * Finds a user by their unique username.
     *
     * @param username The username.
     * @return An Optional containing the generic UserProfileDto if found, otherwise empty.
     */
    Optional<UserProfileDto> findUserByUsername(String username);


    /**
     * Updates the profile of the currently authenticated Owner user.
     *
     * @param updateDTO DTO containing the fields to update.
     * @return The {@link OwnerProfileUpdateResponseDto} containing the updated profile and potentially a new JWT.
     * @throws IOException if image processing fails.
     */
    OwnerProfileUpdateResponseDto updateCurrentOwnerProfile(OwnerProfileUpdateDto updateDTO, @Nullable MultipartFile imageFile )throws IOException;

    /**
     * Updates the profile of the currently authenticated ClinicStaff user.
     * (Note: Admins updating OTHER staff might be in ClinicStaffService).
     *
     * @param updateDTO DTO containing the fields to update.
     * @return The {@link ClinicStaffProfileUpdateResponseDto} containing the updated profile and potentially a new JWT.
     * @throws IOException if image processing fails.
     */
    ClinicStaffProfileUpdateResponseDto updateCurrentClinicStaffProfile(UserProfileUpdateDto updateDTO, @Nullable MultipartFile imageFile ) throws IOException;
}

File: frontend/src/services/userService.ts
import axios from 'axios';
import { API_BASE_URL } from '../config';
import {
    OwnerProfileUpdateDto,
    UserProfileUpdateDto,
    OwnerProfileUpdateResponseDto,      
    ClinicStaffProfileUpdateResponseDto, 
    ApiErrorResponse
} from '../types/apiTypes'; 

/**
 * Updates the currently authenticated owner's profile.
 * Handles sending data as multipart/form-data.
 *
 * @param {string} token - The JWT authentication token.
 * @param {OwnerProfileUpdateDto} updateDTO - The data transfer object for the owner update.
 * @param {File | null} imageFile - The optional new avatar image file.
 * @returns {Promise<OwnerProfileUpdateResponseDto>} A promise resolving to the update response (profile + new token).
 * @throws {Error} Throws an error if the update fails.
 */
export const updateCurrentOwnerProfile = async (
    token: string,
    updateDTO: OwnerProfileUpdateDto,
    imageFile: File | null
): Promise<OwnerProfileUpdateResponseDto> => {
    if (!token) { throw new Error("Authentication token required."); }

    const formData = new FormData();
    const dtoBlob = new Blob([JSON.stringify(updateDTO)], { type: 'application/json' });
    formData.append('dto', dtoBlob);
    if (imageFile) {
        formData.append('imageFile', imageFile, imageFile.name);
    }

    try {
        const response = await axios.put<OwnerProfileUpdateResponseDto>(`${API_BASE_URL}/users/me`, formData, {
            headers: {
                'Authorization': `Bearer ${token}`,
            }
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error('API Update Owner Profile Error:', apiError);
             let errorMessage = 'Failed to update profile.';
             if (typeof apiError.message === 'string') { errorMessage = apiError.message; }
             else if (typeof apiError.message === 'object' && apiError.message !== null) { errorMessage = Object.values(apiError.message).join(' '); }
             else if (apiError.error) { errorMessage = apiError.error; }
             if (error.response.status === 409) errorMessage = `Update failed: ${errorMessage}`; 
            throw new Error(errorMessage);
        } else {
            console.error('Network or unexpected update owner profile error:', error);
            throw new Error('Failed to update profile due to network or unexpected error.');
        }
    }
};

/**
 * Updates the currently authenticated clinic staff's common profile information (username/avatar).
 * Handles sending data as multipart/form-data.
 *
 * @param {string} token - The JWT authentication token.
 * @param {UserProfileUpdateDto} updateDTO - The data transfer object for the staff update (username).
 * @param {File | null} imageFile - The optional new avatar image file.
 * @returns {Promise<ClinicStaffProfileUpdateResponseDto>}  A promise resolving to the update response (profile + new token).
 * @throws {Error} Throws an error if the update fails.
 */
export const updateCurrentClinicStaffProfile = async (
    token: string,
    updateDTO: UserProfileUpdateDto,
    imageFile: File | null
): Promise<ClinicStaffProfileUpdateResponseDto> => {
     if (!token) { throw new Error("Authentication token required."); }

     const formData = new FormData();
     const dtoBlob = new Blob([JSON.stringify(updateDTO)], { type: 'application/json' });
     formData.append('dto', dtoBlob);

     if (imageFile) {
         formData.append('imageFile', imageFile, imageFile.name);
     }

     try {
         const response = await axios.put<ClinicStaffProfileUpdateResponseDto>(`${API_BASE_URL}/users/me/staff`, formData, {
             headers: {
                 'Authorization': `Bearer ${token}`,
             }
         });
         return response.data;
     } catch (error) {
         if (axios.isAxiosError(error) && error.response) {
              const apiError = error.response.data as ApiErrorResponse;
              console.error('API Update Staff Profile Error:', apiError);
             let errorMessage = 'Failed to update profile.';
              if (typeof apiError.message === 'string') { errorMessage = apiError.message; }
              else if (typeof apiError.message === 'object' && apiError.message !== null) { errorMessage = Object.values(apiError.message).join(' '); }
              else if (apiError.error) { errorMessage = apiError.error; }
               if (error.response.status === 409) errorMessage = `Update failed: ${errorMessage}`;
              throw new Error(errorMessage);
          } else {
              console.error('Network or unexpected update staff profile error:', error);
              throw new Error('Failed to update profile due to network or unexpected error.');
          }
     }
};
